doc/design/context.md

```diff
@@ -0,0 +1,175 @@
+## Context Design
+
+`Net` is the container and controller of a set of `Operator`. Each `Operator` in `Net` has a method called `Run`, which will take `Variable` (containing `Tensor`) from `Scope` to make computation. There will be single or several threads to execute operator's `Run` method.
+
+Since the Tensor computation are executed by Eigen library, which needs an Eigen::GpuDevice type object as  parameter in a GPU card, and the GpuDevice parameter is constructed with an Eigen::CudaStreamDevice object. We need to set both a specific GpuID and CudaStream to create a Eigen::CudaStream object.
+
+At the same time, some computation work will executed by cuBLAS or cuDNN library. Take cuBLAS library as an example, we have to acquire a cublasHandle which binds on a CudaStream to make computation. It's the same way as Eigen library does.
+
+A `Context` is corresponded to a thread and holds runtime resources, such as CudaStream, cublasHandle and so on. And the `Run` method of `Operator` will take `Context` from a specific thread as a parameter to get necessary runtime resources.
+
+The `Run` method of `Operator` is defined as follows:
+
+```c++
+Error Run(Scope* scope, Context* context);
+```
+
+
+
+The future DAGNet is run by multi-threads, and each thread will have its own Eigen::GpuDevice object binding on different CudaStream, so that multi-threads can run parallelly on a same GPU card.
+
+And a specific thread will be in charge of copy(communication) work. The copy thread will only need to get a CudaStream from corresponding Context to make data copy between CPU/GPU or GPUs.
+
+We can make a summary:
+
+- Differnet GPU cards have different GpuID, and we can make data parallelism on multi-GPUs.
+- Multi-threads can run a Net parallelly on a single GPU card, and each thread has one Context.
+- There also exists one single thread executing a Net sequentially, and all computation and communication work will use the same Context.
+
+
+Context is defined as follows(just using for unify class CUDAContext and class CPUContext):
+
+```
+class Context {};
+```
+
+### CUDAContext
+
+CUDAContext is defined as follows:：
+
+
+```c++                                   
+class DeviceGuard {
+ public:
+  explicit DeviceGuard(int newDevice)
```

Comment:
Should DeviceGuard take `Place` as the argument? Like

```
class GPUPlaceGuard {
public:
  explicit GPUPlaceGuard(GPUPlace place);
};
```

It could make our design consistent.